File: macos/Tests/ElevateCoreTests/ArmScopeTests.swift
import Testing
@testable import ElevateCore

/// Port of the C# `ArmScopeTests`.
struct ArmScopeTests {
    let mg = "/providers/Microsoft.Management/managementGroups/platform"
    let sub = "/subscriptions/11111111-1111-1111-1111-111111111111"
    var rg: String { sub + "/resourceGroups/prod-rg" }
    var vm: String { rg + "/providers/Microsoft.Compute/virtualMachines/web01" }

    @Test func segmentsReadsSubscriptionResourceGroupAndResource() {
        let segments = ArmScope.segments(vm)
        #expect(segments.map(\.kind) == [.subscription, .resourceGroup, .resource])
        #expect(segments.map(\.name) == ["11111111-1111-1111-1111-111111111111", "prod-rg", "web01"])
        #expect(segments.map(\.scope) == [sub, rg, vm])
    }

    @Test func segmentsReadsAManagementGroupAsOneStep() {
        let segments = ArmScope.segments(mg)
        #expect(segments.count == 1)
        #expect(segments.first == ArmScopeSegment(kind: .managementGroup, name: "platform", scope: mg))
    }

    @Test func segmentsReadsAResourceDirectlyUnderASubscription() {
        let segments = ArmScope.segments(sub + "/providers/Microsoft.Network/virtualNetworks/hub")
        #expect(segments.map(\.kind) == [.subscription, .resource])
    }

    @Test func segmentsReadsChildResourcesAsStepsOfTheirOwn() {
        let scope = vm + "/extensions/monitor"
        let segments = ArmScope.segments(scope)
        #expect(segments.map(\.name) == ["11111111-1111-1111-1111-111111111111", "prod-rg", "web01", "monitor"])
        #expect(segments.last?.scope == scope)
    }

    @Test func segmentsIsEmptyForTheRoot() {
        for scope: String? in [nil, "", "/", "   "] {
            #expect(ArmScope.segments(scope).isEmpty, "scope \(scope ?? "nil")")
        }
    }

    @Test func segmentsKeepsAnUnfamiliarPathAsOneStepRatherThanDroppingIt() {
        let segments = ArmScope.segments("/tenants/contoso/widgets/one")
        #expect(segments.count == 1)
        #expect(segments.first?.kind == .unknown)
    }

    @Test func segmentsIgnoresCaseInThePathKeywords() {
        #expect(ArmScope.segments("/SUBSCRIPTIONS/abc/RESOURCEGROUPS/rg").map(\.kind) == [.subscription, .resourceGroup])
    }

    @Test func isAtOrUnderComparesStepByStep() {
        #expect(ArmScope.isAtOrUnder(vm, ancestor: sub))
        #expect(ArmScope.isAtOrUnder(vm, ancestor: rg))
        #expect(ArmScope.isAtOrUnder(vm, ancestor: vm))
        #expect(ArmScope.isAtOrUnder(vm, ancestor: "/"))
        #expect(!ArmScope.isAtOrUnder(rg, ancestor: vm))
        #expect(!ArmScope.isAtOrUnder(sub, ancestor: rg))
        #expect(!ArmScope.isAtOrUnder(mg, ancestor: sub))
    }

    @Test func isAtOrUnderDoesNotTreatANamePrefixAsAnAncestor() {
        // "/subscriptions/abc" must not swallow "/subscriptions/abcdef".
        #expect(!ArmScope.isAtOrUnder("/subscriptions/abcdef", ancestor: "/subscriptions/abc"))
    }

    @Test func isAtOrUnderIgnoresCase() {
        #expect(ArmScope.isAtOrUnder(vm, ancestor: rg.uppercased()))
    }

    @Test func hasSegmentNamedMatchesAStepsNameOrItsWholeScope() {
        #expect(ArmScope.hasSegmentNamed(vm, name: "prod-rg"))
        #expect(ArmScope.hasSegmentNamed(vm, name: "PROD-RG"))
        #expect(ArmScope.hasSegmentNamed(vm, name: "web01"))
        #expect(ArmScope.hasSegmentNamed(vm, name: rg))
        #expect(ArmScope.hasSegmentNamed(vm, name: rg + "/"))
        #expect(ArmScope.hasSegmentNamed(mg, name: "platform"))
        // A partial name is not a step.
        #expect(!ArmScope.hasSegmentNamed(vm, name: "prod"))
        #expect(!ArmScope.hasSegmentNamed(vm, name: ""))
    }

    @Test func matchesPatternLetsStarCrossSlashes() {
        #expect(ArmScope.matches(pattern: "/subscriptions/*", text: sub))
        #expect(ArmScope.matches(pattern: "/subscriptions/*", text: rg))
        #expect(ArmScope.matches(pattern: "/subscriptions/*", text: vm))
        #expect(!ArmScope.matches(pattern: "/subscriptions/*", text: mg))
        #expect(ArmScope.matches(pattern: "*/resourceGroups/prod-*", text: vm))
        #expect(!ArmScope.matches(pattern: "*/resourceGroups/prod-*", text: sub))
        #expect(ArmScope.matches(pattern: "*managementGroups*", text: mg))
    }

    @Test func matchesPatternMatchesTheWholeStringSoAPrefixAloneIsNotEnough() {
        #expect(!ArmScope.matches(pattern: "/subscriptions", text: sub))
    }

    @Test func looksLikePatternIsAboutTheStar() {
        #expect(ArmScope.looksLikePattern("prod*"))
        #expect(ArmScope.looksLikePattern("*"))
        #expect(!ArmScope.looksLikePattern("prod"))
        #expect(!ArmScope.looksLikePattern(nil))
    }

    @Test func displayNameDropsTheKindTheCaptionAppends() {
        #expect(ArmScope.displayName(detail: "Pay-As-You-Go · subscription") == "Pay-As-You-Go")
        #expect(ArmScope.displayName(detail: "prod-rg · resource group") == "prod-rg")
        #expect(ArmScope.displayName(detail: "just a name") == "just a name")
        #expect(ArmScope.displayName(detail: "") == nil)
        #expect(ArmScope.displayName(detail: nil) == nil)
    }

    @Test func tailKeepsTheLeafAndDropsWhatIsAbove() {
        #expect(ArmScope.tail(vm) == "…/prod-rg/web01")
        #expect(ArmScope.tail(vm, steps: 1) == "…/web01")
        #expect(ArmScope.tail(sub) == sub)
        #expect(ArmScope.tail("/") == "/")
    }
}

File: windows/tests/Elevate.Core.Tests/ArmScopeTests.cs
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class ArmScopeTests
{
    private const string Mg = "/providers/Microsoft.Management/managementGroups/platform";
    private const string Sub = "/subscriptions/11111111-1111-1111-1111-111111111111";
    private const string Rg = Sub + "/resourceGroups/prod-rg";
    private const string Vm = Rg + "/providers/Microsoft.Compute/virtualMachines/web01";

    [Fact]
    public void Segments_ReadsSubscriptionResourceGroupAndResource()
    {
        var segments = ArmScope.Segments(Vm);
        segments.Select(s => s.Kind).Should().Equal(
            ArmScopeKind.Subscription, ArmScopeKind.ResourceGroup, ArmScopeKind.Resource);
        segments.Select(s => s.Name).Should().Equal("11111111-1111-1111-1111-111111111111", "prod-rg", "web01");
        segments.Select(s => s.Scope).Should().Equal(Sub, Rg, Vm);
    }

    [Fact]
    public void Segments_ReadsAManagementGroupAsOneStep()
    {
        var segments = ArmScope.Segments(Mg);
        segments.Should().ContainSingle();
        segments[0].Should().Be(new ArmScopeSegment(ArmScopeKind.ManagementGroup, "platform", Mg));
    }

    [Fact]
    public void Segments_ReadsAResourceDirectlyUnderASubscription()
    {
        var scope = Sub + "/providers/Microsoft.Network/virtualNetworks/hub";
        ArmScope.Segments(scope).Select(s => s.Kind).Should().Equal(ArmScopeKind.Subscription, ArmScopeKind.Resource);
    }

    [Fact]
    public void Segments_ReadsChildResourcesAsStepsOfTheirOwn()
    {
        var scope = Vm + "/extensions/monitor";
        var segments = ArmScope.Segments(scope);
        segments.Select(s => s.Name).Should().Equal("11111111-1111-1111-1111-111111111111", "prod-rg", "web01", "monitor");
        segments[^1].Scope.Should().Be(scope);
    }

    [Theory]
    [InlineData(null)]
    [InlineData("")]
    [InlineData("/")]
    public void Segments_IsEmptyForTheRoot(string? scope) => ArmScope.Segments(scope).Should().BeEmpty();

    [Fact]
    public void Segments_KeepsAnUnfamiliarPathAsOneStepRatherThanDroppingIt()
    {
        var segments = ArmScope.Segments("/tenants/contoso/widgets/one");
        segments.Should().ContainSingle();
        segments[0].Kind.Should().Be(ArmScopeKind.Unknown);
    }

    [Fact]
    public void Segments_IgnoresCaseInThePathKeywords()
    {
        var segments = ArmScope.Segments("/SUBSCRIPTIONS/abc/RESOURCEGROUPS/rg");
        segments.Select(s => s.Kind).Should().Equal(ArmScopeKind.Subscription, ArmScopeKind.ResourceGroup);
    }

    [Theory]
    [InlineData(Vm, Sub, true)]
    [InlineData(Vm, Rg, true)]
    [InlineData(Vm, Vm, true)]
    [InlineData(Rg, Vm, false)]
    [InlineData(Sub, Rg, false)]
    [InlineData(Mg, Sub, false)]
    [InlineData(Vm, "/", true)]
    public void IsAtOrUnder_ComparesStepByStep(string scope, string ancestor, bool expected) =>
        ArmScope.IsAtOrUnder(scope, ancestor).Should().Be(expected);

    [Fact]
    public void IsAtOrUnder_DoesNotTreatANamePrefixAsAnAncestor()
    {
        // "/subscriptions/abc" must not swallow "/subscriptions/abcdef".
        ArmScope.IsAtOrUnder("/subscriptions/abcdef", "/subscriptions/abc").Should().BeFalse();
    }

    [Fact]
    public void IsAtOrUnder_IgnoresCase() =>
        ArmScope.IsAtOrUnder(Vm, Rg.ToUpperInvariant()).Should().BeTrue();

    [Theory]
    [InlineData(Vm, "prod-rg", true)]
    [InlineData(Vm, "PROD-RG", true)]
    [InlineData(Vm, "web01", true)]
    [InlineData(Vm, "prod", false)]
    [InlineData(Vm, Rg, true)]
    [InlineData(Vm, Rg + "/", true)]
    [InlineData(Mg, "platform", true)]
    [InlineData(Vm, "", false)]
    public void HasSegmentNamed_MatchesAStepsNameOrItsWholeScope(string scope, string name, bool expected) =>
        ArmScope.HasSegmentNamed(scope, name).Should().Be(expected);

    [Theory]
    [InlineData("/subscriptions/*", Sub, true)]
    [InlineData("/subscriptions/*", Rg, true)]
    [InlineData("/subscriptions/*", Vm, true)]
    [InlineData("/subscriptions/*", Mg, false)]
    [InlineData("*/resourceGroups/prod-*", Vm, true)]
    [InlineData("*/resourceGroups/prod-*", Sub, false)]
    [InlineData("*managementGroups*", Mg, true)]
    public void MatchesPattern_LetsStarCrossSlashes(string pattern, string scope, bool expected) =>
        ArmScope.MatchesPattern(pattern, scope).Should().Be(expected);

    [Fact]
    public void MatchesPattern_MatchesTheWholeStringSoAPrefixAloneIsNotEnough() =>
        ArmScope.MatchesPattern("/subscriptions", Sub).Should().BeFalse();

    [Theory]
    [InlineData("prod*", true)]
    [InlineData("*", true)]
    [InlineData("prod", false)]
    [InlineData(null, false)]
    public void LooksLikePattern_IsAboutTheStar(string? term, bool expected) =>
        ArmScope.LooksLikePattern(term).Should().Be(expected);

    [Theory]
    [InlineData("Pay-As-You-Go · subscription", "Pay-As-You-Go")]
    [InlineData("prod-rg · resource group", "prod-rg")]
    [InlineData("just a name", "just a name")]
    [InlineData("", null)]
    [InlineData(null, null)]
    public void DisplayName_DropsTheKindTheCaptionAppends(string? detail, string? expected) =>
        ArmScope.DisplayName(detail).Should().Be(expected);

    [Fact]
    public void Tail_KeepsTheLeafAndDropsWhatIsAbove()
    {
        ArmScope.Tail(Vm).Should().Be("…/prod-rg/web01");
        ArmScope.Tail(Vm, 1).Should().Be("…/web01");
        ArmScope.Tail(Sub).Should().Be(Sub);
        ArmScope.Tail("/").Should().Be("/");
    }
}
